### PR TITLE
Add "Offline-first" architecture cookbook recipe

### PR DESCRIPTION
**Pull request by miquelbeltran**
_Description of what this PR is changing or adding, and why:_

Following PRs #11410 and #11394 this PR adds the "Offline-first" cookbook recipe for the Architecture design patterns.

_Issues fixed by this PR (if any):_

Part of https://github.com/flutter/website/issues/11374

_PRs or commits this PR depends on (if any):_

## Presubmit checklist

- [ ] This PR is marked as draft with an explanation if not meant to land until a future stable release.
- [x] This PR doesn’t contain automatically generated corrections (Grammarly or similar).
- [x] This PR follows the [Google Developer Documentation Style Guidelines](https://developers.google.com/style) — for example, it doesn’t use _i.e._ or _e.g._, and it avoids _I_ and _we_ (first person).
- [x] This PR uses [semantic line breaks](https://github.com/dart-lang/site-shared/blob/main/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks) of 80 characters or fewer.

### CODE REVIEW

**Review by ericwindmill — APPROVED** [review]

File: src/content/cookbook/architecture/offline-first.md [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+
+A different approach for synchronization 
+is to use a push service to provide up-to-date data to the application. 
+In this case, the server notifies the application when data has changed, 
+instead of being the application asking for updates.
+
+For example, using [Firebase messaging][], 
+you can push small payloads of data to the device, 
+and as well, 
+you can trigger synchronization tasks remotely using background messages.
```
Comment (sfshaza2) [inline src/content/cookbook/architecture/offline-first.md:494]:
```suggestion
as well as trigger synchronization tasks remotely using background messages.
```

File: src/content/cookbook/architecture/offline-first.md [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+when the stored data needs to be updated via a push notification.
+
+You can combine both approaches together, 
+having a background synchronization task and using background push messages, 
+to keep the application database synchronized with the server.
+
+## Putting it all together
+
+Writing an offline-first application 
+requires making different decisions regarding
```
Comment (sfshaza2) [inline src/content/cookbook/architecture/offline-first.md:507]:
```suggestion
requires making decisions regarding 
```

OR change "different" to "several".

**Review by sfshaza2 — COMMENTED** [review]
Some rephrasing.

File: examples/cookbook/architecture/offline_first/lib/domain/model/user_profile.freezed.dart [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+part of 'user_profile.dart';
+
+// **************************************************************************
+// FreezedGenerator
+// **************************************************************************
+
+T _$identity<T>(T value) => value;
+
+final _privateConstructorUsedError = UnsupportedError(
+    'It seems like you constructed your class using `MyClass._()`. This constructor is only meant to be used by freezed and you are not supposed to need it nor use it.\nPlease check the documentation here for more information: https://github.com/rrousselGit/freezed#adding-getters-and-methods-to-our-models');
```
[the remark this answers is not in the document]
Comment (miquelbeltran) [reply examples/cookbook/architecture/offline_first/lib/domain/model/user_profile.freezed.dart:15]: Hi @sfshaza2 this is a generated file by the _freezed_ package, I'd not change it.

File: src/content/cookbook/architecture/offline-first.md [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+    required String name,
+    required String photoUrl,
+  }) = _UserProfile;
+}
+```
+
+In apps that have complex data, 
+for example when the remote data contains more fields than the needed by the UI,
+you may want to have a different data class for the API and database services,
+and one for the UI.
```
[the remark this answers is not in the document]
Comment (miquelbeltran) [reply src/content/cookbook/architecture/offline-first.md:105]: > to have a one data class

Should probably become

> to have one data class

File: src/content/cookbook/architecture/offline-first.md [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+```
+
+Your synchronization logic should attempt
+to send it to the API service 
+only when the `synchronized` flag is `false`.
+If the request is successful, then change it to `true`.
+
+### Pushing data from server
+
+A different approach for synchronization
```
[the remark this answers is not in the document]
Comment (miquelbeltran) [reply src/content/cookbook/architecture/offline-first.md:486]: Yes, it is explained later on that devs can combine both approaches, and are not mutually exclusive.

[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/content/cookbook/architecture/offline-first.md:483]: Yes, I saw that later. :)

**Review by sfshaza2 — APPROVED** [review]
lgtm!

**Pull request merged**